core/src/main/java/com/graphhopper/search/EdgeKVStorage.java

```diff
@@ -30,37 +30,50 @@
  * @author Peter Karich
  */
 public class EdgeKVStorage {
+
     private static final long EMPTY_POINTER = 0, START_POINTER = 1;
     // Store the key index in 2 bytes. Use negative values for marking the value as duplicate.
     static final int MAX_UNIQUE_KEYS = (1 << 15);
     // Store string value as byte array and store the length into 1 byte
     private static final int MAX_LENGTH = (1 << 8) - 1;
-    boolean throwExceptionIfTooLong = false;
     private final DataAccess keys;
-    // storage layout per entry:
-    // 1 byte    | 2 bytes  | 1 byte      | x    | 2 bytes  | 1 byte      | x    | 2 bytes  (dup example) | 4 bytes | ...
-    // vals count| key_idx_0| val_length_0| val_0| key_idx_1| val_length_1| val_1| -key_idx_2             | delta_2 | key_idx_3 | val_length_3 | val_3
-    // Drawback: we need to loop through the entries to get the start of val_x.
-    // Note, that we detect duplicate values via smallCache and then use the negative key index as 'duplicate' marker.
-    // We then store only the delta (signed int) instead the absolute unsigned long value to reduce memory usage when duplicate entries.
+    // storage layout per entry, examples:
+    // vals count (1 byte)|
+    // key_idx_0  (2 byte)| val_length_0 (1 byte)| val_0 (x bytes) <- store String or byte[] with dynamic length
+    // key_idx_1  (2 byte)| int          (4 byte)                  <- store int with fixed length
+    // -key_idx_2 (2 byte)| delta_2      (4 byte)                  <- store duplicate, then key is negative
+    // key_idx_3  (2 byte)| val_length_3 (1 byte)| val_3 (x bytes)
```

Comment:
I am not sure I understand this. Is this the layout of the `vals` DataAccess? Or the one of the `keys` DataAccess, or maybe a combination of both? Can you maybe add some explanation of the general idea how data is stored here?

I guess I understand the first and the fourth lines: We use key_idx_0/3 to point into the vals DataAccess and the first byte we read tells us the length of the value in bytes x which we can then read. But what about the second line? It says we just read four bytes as an int, but how do we know the first byte is not the length as for the other two? And for the third line we see a negative index that is used for duplicates. But why do we not just point to the same value for duplicates? I.e. when key_idx_0 and key_idx_2 are supposed to point to the same value why don't they just store the same key (which points to the value they are both referring to)?

Sorry I am probably missing some basics here, but then again if this documentation is supposed to explain how the index worked I feel like it isn't telling me enough :)

Reply:
Ok just to say it in my own words: We have a `vals` array that, for each edge, stores the number of keys used for this edge followed by triples `key/value_length/value`. For every edge we store a pointer into this array pointing to the first key/value pair of that edge in the edge flags (`E_NAME`). This way, given an edge, we can use the pointer to find the first key/value pair and traverse all the key/value pairs in order by jumping to the next using the value_length. We know when to stop, because we also stored the number of keys. Each key is stored as two-byte number (key_idx) and we store the mapping between key_idxs and the actual (String) keys in a separate `keys` array. Besides the key we also store the 'type' of each value and for some types we skip storing the value_length and derive the value_length from the type, i.e. 4 bytes for an int and so on. We prevent storing the same value multiple times by storing a pointer to the first occurrence of a value instead of that value. We recognize that a value is a pointer to the actual value by negating the key, i.e. key=-6 means key=6, but the value is a pointer to the actual value.